### ISSUE TITLE
Batch norm layer in stax does not compute moving averages for test-time

### ISSUE THREAD

**Issue by alexbw** [google/jax#139]
It should do this.

**Comment by jonasrauber** [comment]
I think this should be labeled as _bug_, not _enhancement_

**Comment by mattjj** [comment]
Isn't it a missing feature? (Forgive me, I don't know much about batch norm so I'm probably misunderstanding! That's why I keep writing bugs into its implementation...)

**Comment by jonasrauber** [comment]
I guess it depends. If stax just wouldn't have a BatchNorm layer, I would agree. But having one that does (in some cases) not behave in the way it should qualifies as a bug to me, especially because the broken case (test-time usage) is highly relevant (even during training you usually do some validation in between). Moreover, there is no loud error produced during test time, it will just behave like during training and people won't notice that it's wrong.

Plus, in some sense it also behaves wrong during training, because the computation of the moving averages (that are then used during test-time) should happen there (during training). So one cannot use the stax BatchNorm layer for training and then export the trained model to some other framework that has a correct test-time implementation of BatchNorm.

**Comment by jonasrauber** [comment]
Adding to this: the current implementation also breaks if the `axis` argument takes some other values, e.g. `axis=(0, 2, 3)`, where the missing axis is not the last one, because the `axis` are dropped for `gamma` and `beta` and not restored later.
I guess this can be fixed by removing this line: https://github.com/google/jax/blob/7e93bff2b95e6f49be7e2cd310e8953bbbdcdc9d/jax/experimental/stax.py#L134

**Comment by mattjj** [comment]
Thanks for explaining, you're absolutely right!

Would you mind opening another issue for the `axis` arg bug?

**Comment by jonasrauber** [comment]
done, see #461

**Comment by hawkinsp** [comment]
We're not planning on working further on the current iteration of stax, in favor of supporting other libraries such as Flax and Haiku. Closing.
